### ISSUE TITLE
Error message when updating card info

### ISSUE THREAD

**Issue by IowaBoy1** [iATSPayments/com.iatspayments.civicrm#264]
Drupal 7.61
Civi 5.4.1
iATS 1.6.2

When donors change their credit card info and SAVE it, they get multiple popups (1) a "Authorization code" Click OK, (2) "The Recurring contribution billing info was successfully updated." Click OK (3) "Cannot connect to the network server."

It does process the card information and store the new info. But to the donor, it appears that it failed.

I do not see any errors in the Civi Logs

I do have errors in the Drupal Logs. Below are the 7 related logs to the most recent card change--including the login.

access denied | 11/24/2018 - 23:02 | civicrm | UserName |  
  | php | 11/24/2018 - 23:02 | Strict warning: Only variables should be assigned by... | UserName |  
  | php | 11/24/2018 - 23:02 | Warning: Invalid argument supplied for foreach() in... | UserName |  
  | php | 11/24/2018 - 23:02 | Notice: Undefined property: CRM_Contribute_Form... | UserName |  
  | php | 11/24/2018 - 23:02 | Warning: Invalid argument supplied for foreach() in... | UserName |  
  | php | 11/24/2018 - 23:02 | Notice: Undefined property: CRM_Contribute_Form... | UserName |  
  | user | 11/24/2018 - 22:59 | Session opened for UserName | UserName

The Strict Warning reads:
MESSAGE | Strict warning: Only variables should be assigned by reference in CRM_Contribute_Form_UpdateBilling->postProcess() (line 257 of ......./modules/civicrm/CRM/Contribute/Form/UpdateBilling.php).
-- | --
SEVERITY debug

The Access Denied reads:
LOCATION | https://secure.radiusinternational.org/civicrm/user&force=1?snippet=json
-- | --
REFERRER https://secure.radiusinternational.org/civicrm/user
MESSAGE civicrm
SEVERITY warning

**Comment by adixon** [comment]
Fascinating, definitely worth pursuing. I had a similar but less drastic experience updating my credit card as a donor (i.e not as an admin), so I'm pretty sure it's some kind of ajax-content permission issue. It may be iATS code related, but I'm not convinced ...

**Comment by IowaBoy1** [comment]
So if is some kind of ajax-content permission issue, where/how do I start to look for the solution?

**Comment by adixon** [comment]
Yeah, those are tricky - sometimes you can look in your browser network inspect box and see what the url was the triggered the error. You can also grep the code base for the specific text that appears in the boxes. You could also try posting your experience on the stackexchange and see if anyone there has more clues.

**Comment by adixon** [comment]
Here's another piece of the puzzle: there's an outstanding patch to civicrm core required in order to restore this function properly in CiviCRM 5:
https://lab.civicrm.org/dev/core/issues/332

It's a very small patch, if you're able to confirm it resolves your problem then maybe we can get some more momentum behind it.

**Comment by guyiac** [comment]
I could do that. Or I could just recommend to the client that they use Stripe.

**Comment by KarinG** [comment]
Hi - I’ve posted the issue here: https://civicrm.stackexchange.com/questions/28450/why-can-users-no-longer-update-their-own-billing-credit-card-details-using-iats - it’s a CiviCRM Core regression that we have a one line of code fix for (see details on that SE ticket). It affects all payment processors - though we may have been one of the only ones using that CiviCRM Core feature.

**Comment by IowaBoy1** [comment]
Thanks guys! I have updated that line of code. And I no longer get the message "Cannot reach server" when clicking Change Billing. However, I still get the error if I click View or Edit. Any ides on that?

**Comment by KarinG** [comment]
Glad that partially does the trick - please feel free to add a comment to that PR that even though it got closed it helped resolve the regression for you. 

Send me screenshots/steps to reproduce the View and Edit 403 errors and I’ll have a stab at it.

**Comment by IowaBoy1** [comment]
![screenshot 2019-02-13 15 15 19](https://user-images.githubusercontent.com/28285267/52750980-b89cc500-2fa2-11e9-9882-1501c6ef81c7.png)

When logged in as a donor and in the donor dashboard. At Recurring contributions if I click on View or Edit I get the following error. I no longer get this error when I click Change Billing Info

![screenshot 2019-02-13 15 13 59](https://user-images.githubusercontent.com/28285267/52750994-c94d3b00-2fa2-11e9-8090-40f163fb9bc3.png)
![screenshot 2019-02-13 15 13 19](https://user-images.githubusercontent.com/28285267/52751004-d0744900-2fa2-11e9-8fdf-cd29584f8bc2.png)

**Comment by IowaBoy1** [comment]
Previously I had just clicked on Change BIlling Details and it did not give me a "Unable to Reach server error" immediately like I do with the View/Edit.  Now I tested actually changing the card info and I still get the original 3 messages in my first post on this thread with the final one being "Unable to reach server".

**Comment by IowaBoy1** [comment]
I updated my CRM/iATS/Form/IATSCustomerUpdateBillingInfo.php with #275 But I still have the same error message.

**Comment by adixon** [comment]
Great pictures. I notice that the console is showing a 403 error, which suggests a permissions related error. So I suspect there's something in your configuration that is telling the code to disallow the operation.

I'm not certain about this, but it might help for debugging purposes to turn off the "enable popup forms" setting (here: civicrm/admin/setting/preferences/display?reset=1).

It does seem like some kind of incoherence in the code ...

**Comment by IowaBoy1** [comment]
Thanks Alan, tried it but I still get the "cannot reach server" message.

**Comment by adixon** [comment]
So - I guess that means this form is always accessed via ajax? Hmm ...

I've just recreated a similar but different error - when I'm editing the recurring record (as an unprivileged but authenticated user), get a 403 ajax error on a url of the form:

civicrm/custom?type=ContributionRecur&entityID=26&qf=longstringhere&cgcount=1&snippet=json

**Comment by IowaBoy1** [comment]
So??? I have to change privilege settings? What is an unprivileged but authenticated user?

**Comment by IowaBoy1** [comment]
Here is an interesting thing.
After I turned off the "enable popup forms" setting (here: civicrm/admin/setting/preferences/display?reset=1). Per Alan's suggestion above, the donor would get an error message that the recur id XXX was not valid. The message would increment up one number with each time. However, it was not recording the contribution in the contribution or the contribution recur tables. It did process in iATS.

What finally fixed the issue was when I turned back on the "enable popup forms".

Here is the Drupal log error for this 
The CiviCRM "contribution transact" API returned the error: "contribution_recur_id is not valid : 515" when called by function "submitLivePayment" on line 1759 of wf_crm_webform_postprocess.inc with parameters: "Array ( [credit_card_type] =>....

And here is the CiviCRM Log for it
[info] $Fatal Error Details = Array
(
    [callback] => Array
        (
            [0] => CRM_Core_Error
            [1] => exceptionHandler
        )

    [code] => -3
    [message] => DB Error: constraint violation
    [mode] => 16
    [debug_info] => INSERT INTO civicrm_contribution (contact_id , financial_type_id , contribution_page_id , payment_instrument_id , receive_date , non_deductible_amount , total_amount , fee_amount , net_amount , trxn_id , invoice_id , invoice_number , currency , source , contribution_recur_id , is_test , contribution_status_id , check_number , tax_amount ) VALUES ( 3 ,  2 ,  4 ,  1 ,  20190222071225 ,  10 ,  10 ,  0 ,  10 , 'AB74A542:1550848347' , '233e0296dc857a0eea413160ebae521d' , 'INV_19664' , 'USD' , 'Pay Tuition' ,  516 ,  0 ,  1 ,  NULL ,  0 )  [nativecode=1452 ** Cannot add or update a child row: a foreign key constraint fails (`radius_2_civicrm`.`civicrm_contribution`, CONSTRAINT `FK_civicrm_contribution_contribution_recur_id` FOREIGN KEY (`contribution_recur_id`) REFERENCES `civicrm_contribution_recur` (`id`) ON DE)]
    [type] => DB_Error
    [user_info] => INSERT INTO civicrm_contribution (contact_id , financial_type_id , contribution_page_id , payment_instrument_id , receive_date , non_deductible_amount , total_amount , fee_amount , net_amount , trxn_id , invoice_id , invoice_number , currency , source , contribution_recur_id , is_test , contribution_status_id , check_number , tax_amount ) VALUES ( 3 ,  2 ,  4 ,  1 ,  20190222071225 ,  10 ,  10 ,  0 ,  10 , 'AB74A542:1550848347' , '233e0296dc857a0eea413160ebae521d' , 'INV_19664' , 'USD' , 'Pay Tuition' ,  516 ,  0 ,  1 ,  NULL ,  0 )  [nativecode=1452 ** Cannot add or update a child row: a foreign key constraint fails (`radius_2_civicrm`.`civicrm_contribution`, CONSTRAINT `FK_civicrm_contribution_contribution_recur_id` FOREIGN KEY (`contribution_recur_id`) REFERENCES `civicrm_contribution_recur` (`id`) ON DE)]
    [to_string] => [db_error: message="DB Error: constraint violation" code=-3 mode=callback callback=CRM_Core_Error::exceptionHandler prefix="" info="INSERT INTO civicrm_contribution (contact_id , financial_type_id , contribution_page_id , payment_instrument_id , receive_date , non_deductible_amount , total_amount , fee_amount , net_amount , trxn_id , invoice_id , invoice_number , currency , source , contribution_recur_id , is_test , contribution_status_id , check_number , tax_amount ) VALUES ( 3 ,  2 ,  4 ,  1 ,  20190222071225 ,  10 ,  10 ,  0 ,  10 , 'AB74A542:1550848347' , '233e0296dc857a0eea413160ebae521d' , 'INV_19664' , 'USD' , 'Pay Tuition' ,  516 ,  0 ,  1 ,  NULL ,  0 )  [nativecode=1452 ** Cannot add or update a child row: a foreign key constraint fails (`radius_2_civicrm`.`civicrm_contribution`, CONSTRAINT `FK_civicrm_contribution_contribution_recur_id` FOREIGN KEY (`contribution_recur_id`) REFERENCES `civicrm_contribution_recur` (`id`) ON DE)]"]
)

**Comment by IowaBoy1** [comment]
I am coming back around to this. I had just turned off the self-help check boxes in the iATS settings in CiviContribute. I have been getting a lot of requests lately for card edits so I decided to try this again and see if it has been fixed. Now I see "More" but it is not a link like it used to be. Has this been abandoned entirely? Or should I be able to have our donors be able to edit their card number and expiry?
